### PR TITLE
Assignment2

### PR DESCRIPTION
**Pull request by Veltoc**
worked with: @MattJett-EWU

As partners go we don't need it to be reviewed

### CODE REVIEW

File: SecretSanta/src/SecretSanta.Data/User.cs [hunk 1]
```diff
@@ -1,24 +1,21 @@
...
-            Id = id;
-            FirstName = firstName;
-            LastName = lastName;
-            Gifts = gifts;
-        }
+#nullable disable
+        public ICollection<Gift> Gifts { get; set; }
+        public List<UserGroup> UserGroups { get; set; }
+#nullable enable
+        // public Nullable<User> Santa { get; set; } nullable property? something may need to be changed to make this work. or its something else
```
Comment (Veltoc) [inline SecretSanta/src/SecretSanta.Data/User.cs]: should this code be uncommented or should it be deleted

File: .gitignore [hunk 2]
```diff
@@ -273,3 +273,4 @@ __pycache__/
 # tools/
 
 # End of https://www.gitignore.io/api/aspnetcore
+/BlogEngine
```
Comment (Veltoc) [inline .gitignore]: whats this for?

File: SecretSanta/GlobalSuppression.cs [hunk 3]
```diff
@@ -1,6 +1,10 @@
+//suppressed warnings from the project, not adding justifification for default
 [assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1056:Uri properties should not be strings", Justification = "<Pending>")]
 [assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1054:Uri parameters should not be strings", Justification = "<Pending>")]
 [assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Reliability", "CA2007:Consider calling ConfigureAwait on the awaited task", Justification = "<Pending>")]
 [assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1052:Static holder types should be Static or NotInheritable", Justification = "<Pending>")]
 [assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Naming", "CA1707:Identifiers should not contain underscores", Justification = "<Pending>")]
-[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Performance", "CA1825:Avoid zero-length array allocations.", Justification = "<Pending>")]
\ No newline at end of file
+[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Performance", "CA1825:Avoid zero-length array allocations.", Justification = "<Pending>")]
```
Comment (Veltoc) [inline SecretSanta/GlobalSuppression.cs]: justification?

**Comment by breaman** [comment]
##  (Veltoc)

## Assignment2

| Project Analysis Summary |   |
| -----------------| - |
| Nullable Enabled: | **✔** |
| Unexpected Disabled Warnings:  | **✔**
| Appropriately handle all warnings: | **❌** |

| ProjectName | Nullable | Solutions | Analyzers | Unexpected Disabled Warnings | Waning Level |
| -- | -- | -- | -- | -- | -- |
| SecretSanta.Api | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Business | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Data | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Web | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer |  | 4 |
| SecretSanta.Api.Tests | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Business.Tests | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Data.Tests | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Web.Tests | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer |  | 4 |


## Main Assignment

1. Add the following classes to the **SecretSanta.Data** project with the following associated properties:
   - **`EntityBase`** ✔
     - `int Id`
   - **`FingerPrintEntityBase`** inherit from **`EntityBase`** ✔
     - `string CreatedBy`
     - `DateTime CreatedOn`
     - `string ModifiedBy`
     - `DateTime ModifiedOn`
   - **`Group`** inherit from **`FingerPrintEntityBase`** ✔
     - `string Name`
   - Create a class that will hold the many to many relationship between User and Group ✔
     - Add a collection on both User and Group that references this class
2. Modify the following classes in the **SecretSanta.Data** project:
   - **`Gift`** ✔
     - Remove the Id property and inherit from FingerPrintEntityBase
   - **`User`** ✔
     - Remove the Id property and inherit from FingerPrintEntityBase
     - Make `Gifts` read/write
     - Create a nullable Santa property that contains a User as someone's Santa
3. Remove non-default constructors for `Gift` and `User`. ✔
4. In the **SecretSanta.Data** project
   - Create the **`ApplicationDbContext`** class ✔
     - Create two constructors ✔
        - One only takes the DbContextOptions and passes it on to the base class
        - One that takes the DbContextOptions and an IHttpContextAccessor, pass the DbContextOptions to the base class and save off the IHttpContextAccessor in a class property
     - Add DbSet properties for User, Gift, and Group classes ✔
     - Override OnModelCreating method and use the fluent API to configure the many-to-many pieces that are needed for setting up the database properly ✔
     - Override both SaveChanges methods and apply the fingerprinting logic ✔
5. Update unit tests in the **`SecretSanta.Data.Tests`** project to test actual database logic instead of just constructor logic
   - Mock up IHttpContextAccessor to verify that fingerprint information is working correctly (using Moq library) ✔


## Extra Credit

The following are options for extra credit:
- Enable console logging that will log all database commands ✔

**Pull request closed without merging**